src/D2L.CodeStyle.Analyzers/Diagnostics.cs

```diff
@@ -514,5 +514,77 @@ public static class Diagnostics {
 			isEnabledByDefault: true,
 			description: "We require that event types be sealed."
 		);
+
+		public static readonly DiagnosticDescriptor MutableBaseType = new DiagnosticDescriptor(
+			id: "D2L0061",
+			title: "Base class for immutable type must be [Immutable] (or [ImmutableBaseClass])",
+			messageFormat: "{0}'s base class is {1} which is missing [Immutable] (or, more weakly, [ImmutableBaseClass]).",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor ArraysAreMutable = new DiagnosticDescriptor(
+			id: "D2L0062",
+			title: "Arrays are mutable and thus can't be held by an immutable type.",
+			messageFormat: "{0}[]'s are mutable even if {0}'s are immutable because array items can be substituted. Prefer an immutable container type such as ImmutableArray<{0}>.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor DynamicObjectsAreMutable = new DiagnosticDescriptor(
+			id: "D2L0063",
+			title: "Dynamic objects are mutable and thus can't be held by an immutable type.",
+			messageFormat: "Dynamic objects are mutable and thus can't be held by an immutable type.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor TypeParameterIsNotKnownToBeImmutable = new DiagnosticDescriptor(
+			id: "D2L0064",
+			title: "Type parameter is not known to be immutable.",
+			messageFormat: "The type parameter {0} is not known to be immutable. Add [Immutable] to it.",
```

Comment:
"The type parameter T is not known to be immutable. Add [Immutable] to it."

Reply:
Given the way we're going with the container analyzer, the wording on this seems off to me.

> Type parameter is not known to be immutable.

This implies to me that if I add `<T>` -> `<[Immutable] T>`, then `T` is guaranteed to be immutable. This aligns with current behaviour of immutable generics.

> Type parameter is not considered to be immutable.

Splitting hairs, but the slightly different wording I think alludes to the real situation a bit better. "This may or may not be immutable, but if you want it to be considered to be immutable for the purpose of analyzing your class declaration, add `[Immutable]` to it."

Reply:
> This aligns with current behaviour of immutable generics.

Yeah, I was kinda trying to not step on your toes :)

Agreed the wording is not fantastic. Want to fix this when you do the container stuff?